blog/content/post/2020-02-19-opnsense-xen-trenchboot-demo.md

```diff
@@ -0,0 +1,171 @@
+Measurements of each are stored in Trusted Platform Module (TPM) special
+registers. After coreboot ends its execution, firmware is being verified and if
+it is not corrupted, next stage of booting starts.
+
+>For details about SRTM I refer to articles and papers available in Internet
+(e.g. [this
+one](http://security.hsr.ch/mse/projects/2011_Root_of_Trust_for_Measurement.pdf))
+
+#### TrenchBoot
+
+TrenchBoot is the next stage running after coreboot execution ended. Security
+concept is exactly the same as in previous case - measure and verify running
+environment. TrenchBoot uses slightly different technology called **Dynamic Root
+of Trust for Measurements** (DRTM). Pointing out differences is not a subject of
+this article. To be honest, it is very complex topic and understanding it could
+take too much time. What you need to know (or trust us about) is that other
+TPM's registers store those measurements and running environment is verified and
+can be described as trusted.
+
+Below diagram presents described technologies and features. It is simplified,
+but should give you main idea about their usage and role.
+
+![coreboot_trenchboot](../../static/img/coreboot_trenchboot.png)
+
+## Virtualization, isolation and WiFi performance
+
+If virtualization, hypervisor, Xen, Dom0 and related terminology is foreign to
+you, please get familiar with those blog posts first:
+
+- [Xen Project](https://blog.3mdeb.com/2020/2020-02-05-meta-pcengines-xen/)
+- [pfSense under Xen - part 1](https://blog.3mdeb.com/2019/2019-11-06-pfsense-under-xen-introduction/)
+- [pfSense under Xen - part 2](https://blog.3mdeb.com/2019/2019-12-13-pfsense-boot-under-xen/)
+
+When you already have basic knowledge, we can move on to the essential part of
+this section.
+
+There are issues with WiFi support on BSD-based firewalls. Many newer cards are
+simply not usable due to lack of specific driver - Compex WLE600VX is a good
+example. Our idea is to run `Xen` with `Dom0` and `OPNSense` in virtual machine. What
+we gain with that approach? Dom0, as privileged domain, has direct access to
+hardware. Also, our custom and adapted build of it, supports WLE600VX WiFi card.
+OPNSense supports popular Ethernet Network Interfaces (e.g. Intel E1000). We
+just fasten it together by passing WLE600VX WiFi from Dom0 to OPNSense as
+emulated Intel E1000 NIC. It is so simple and so effective. For imagination of
+that configuration, look at the diagram below.
+
+![coreboot_trenchboot](../../static/img/wle600-xen-opn.png)
+
+As a result, we have reliable and working WiFi Card in router - although
+OPNSense doesn't directly support it. There is also another benefit achieved
+by virtualization. It introduces another level of security. Closing entities
+(such as network interface, firewall software or operating system) in separate
+environments minimizes the effects of a potential attack. User can be sure that
+virus doesn't spread out to entire system, but stay only in infected part.
```

Comment:
@piotr-kleins it works both ways, firmware from wifi card has hard time to infect system if correct IOMMU configuration was used and firewall VM is isolated from real hardware. The question is what is the performance penalty? We should provide benchmarks to prove our thesis.